### ISSUE TITLE
getwork fail (core dumped)

### ISSUE THREAD

**Issue by hackmod** [ethereum-mining/ethminer#1689]
**Describe the bug**
~~~sh
$ ~/src/ethminer/build/ethminer/ethminer -U --cuda-devices 0 -P http://127.0.0.1:8485
...
 i 16:43:01 cuda-0   **Accepted   1 ms. 127.0.0.1
 i 16:43:01 cuda-0   Solution: 0x958c867362f2b804
 i 16:43:01 cuda-0   **Accepted   1 ms. 127.0.0.1
 i 16:43:01 cuda-0   Solution: 0x958c867362f6354e
 i 16:43:01 cuda-0   **Accepted   0 ms. 127.0.0.1
 i 16:43:01 cuda-0   Solution: 0x958c867362f6c453
 X 16:43:01 getwork  Failed getting work!
 i 16:43:01 main     Disconnected from 127.0.0.1
Segmentation fault (core dumped)
~~~

**To Reproduce**
my `test-genesis.json` (with low difficulty)
~~~
{
  "config": {
    "chainId": 1347,
    "homesteadBlock":0,
    "eip155Block":0,
    "eip158Block":0,
    "progpowBlock":null,
    "ethash": { }
  },
  "difficulty": "0x2000", // "0x20000" also tested
  "timestamp": "0x5be6530c",
  "nonce": "0x0000000000000042",
  "mixHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "parentHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "extraData": "0x",
  "gasLimit": "3141592",
  "alloc": { }
  }
}
~~~
2. init genesis // run `geth`
3. run `ethminer` with getwork parameters

**Expected behavior**
before `0.15.0.dev11` (?) it works nicely, but after some `libpoolprotocols` updates, it broken.

- b0b172fee (0.15.0.dev11) - work
- 4f476af39 (0.16.0.dev0) - not work (segfault)

**Screenshots (Optional)**
![image](https://user-images.githubusercontent.com/32324335/48758783-90276680-ece4-11e8-9a1f-fb70c5842d7f.png)

**Desktop (please complete the following information):**
- Operating System: ubuntu 16
- Hardware P106-6G
- Ethminer Version: above all ethminer v0.16.x

**Comment by AndreaLanfranchi** [comment]
I've recently completed a test work to refactor completely Getwork.
If you're able to build from source could you please test this fork ?
https://github.com/AndreaLanfranchi/ethminer

**Comment by hackmod** [comment]
~~~sh
$ ethminer -U --cuda-devices 0 -P http://127.0.0.1:8485
ethminer 0.17.0.dev0-96+commit.ab4fd644.dirty
Build: linux/release/gnu

cu 18:35:27 ethminer Using grid size: 8,192, block size: 128, streams: 2 parallel hashes : 4
 i 18:35:27 ethminer Configured pool 127.0.0.1:8485
 i 18:35:27 ethminer Selected pool 127.0.0.1:8485
 i 18:35:27 ethminer Established connection to 127.0.0.1:8485
 i 18:35:27 ethminer Spinning up miners...
 X 18:35:27 ethminer 127.0.0.1:8485 reported status 400 Bad Request
 i 18:35:27 ethminer Disconnected from 127.0.0.1:8485
 i 18:35:27 ethminer No connection. Suspend mining ...
 i 18:35:27 ethminer Selected pool 127.0.0.1:8485
 i 18:35:27 ethminer Established connection to 127.0.0.1:8485
 i 18:35:27 ethminer Resume mining ...
 X 18:35:27 ethminer 127.0.0.1:8485 reported status 400 Bad Request
 i 18:35:27 ethminer Disconnected from 127.0.0.1:8485
 i 18:35:27 ethminer No connection. Suspend mining ...
 i 18:35:27 ethminer Selected pool 127.0.0.1:8485
 i 18:35:27 ethminer Established connection to 127.0.0.1:8485
 i 18:35:27 ethminer Resume mining ...
 X 18:35:27 ethminer 127.0.0.1:8485 reported status 400 Bad Request
 i 18:35:27 ethminer Disconnected from 127.0.0.1:8485
 i 18:35:27 ethminer No connection. Suspend mining ...
 i 18:35:27 ethminer No more connections to try. Exiting...
 i 18:35:27 ethminer Shutting down miners...
 i 18:35:27 main     Signal intercepted ...
 i 18:35:27 ethminer Terminated!
~~~

**Comment by AndreaLanfranchi** [comment]
There is something wrong with your Geth node.
It responds with HTTP status 400 to ethminer's requests.

From a command prompt try execute
```
curl -X POST -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"eth_getWork","params":[],"id":1}' http://127.0.0.1:8545
```
and report output

**Comment by hackmod** [comment]
this is my test curl script
```sh
curl -i --header "Content-Type: application/json" -X POST --data '{"worker": "foobar", "jsonrpc":"2.0","id":1,"method":"eth_getWork","params":[]}'
```
and the result
```
{"jsonrpc":"2.0","id":1,"result":["0xfc7ccb8b8afb98baf95299461b82b429bfcb46081a0e1d3145f8215b96de92be","0x0000000000000000000000000000000000000000000000000000000000000000","0x00007fc01ff007fc01ff007fc01ff007fc01ff007fc01ff007fc01ff007fc01f","0x2735"]}
```

**Comment by AndreaLanfranchi** [comment]
> this is my test curl script

Please remove the "worker" member and resubmit.
What is the output ?

**Comment by hackmod** [comment]
~~~sh
NODE=http://127.0.0.1:8485
curl -i --header "Content-Type: application/json" -X POST --data '{"jsonrpc":"2.0","id":1,"method":"eth_getWork","params":[]}' $NODE
~~~

~~~
{"jsonrpc":"2.0","id":1,"result":["0xfc7ccb8b8afb98baf95299461b82b429bfcb46081a0e1d3145f8215b96de92be","0x0000000000000000000000000000000000000000000000000000000000000000","0x00007fc01ff007fc01ff007fc01ff007fc01ff007fc01ff007fc01ff007fc01f","0x2735"]}
~~~

**Comment by AndreaLanfranchi** [comment]
One last try (just to be sure).
Can you run ethminer against dwarfpool geth port ?

Use 
```
-P http://eth-eu.dwarfpool.com:80/<yourethaddress>
```

and tell me what happens ?

**Comment by hackmod** [comment]
![image](https://user-images.githubusercontent.com/32324335/48768021-6cbbe600-ecfb-11e8-860b-4283e365392a.png)

**Comment by hackmod** [comment]
when i append worker option, it does not emit `400` error but..

![image](https://user-images.githubusercontent.com/32324335/48768198-e653d400-ecfb-11e8-9187-caca79c211a0.png)

**Comment by AndreaLanfranchi** [comment]
Ok that makes sense and I know where to patch. (the worker thing).

The second problem is your work notification passes a seed hash all zeroed

```
{
  "jsonrpc": "2.0",
  "id": 1,
  "result": [
    "0xfc7ccb8b8afb98baf95299461b82b429bfcb46081a0e1d3145f8215b96de92be",
    "0x0000000000000000000000000000000000000000000000000000000000000000",
    "0x00007fc01ff007fc01ff007fc01ff007fc01ff007fc01ff007fc01ff007fc01f",
    "0x2735"
  ]
}
```
Second element of the result array member
Thus ethminer is not able to determine epoch and, again, can't build the DAG.
I will manage this unfortunate case in code.

Nevertheless your node **should** output a seed hash

**Comment by AndreaLanfranchi** [comment]
R u trying to implement ProgPoW on ethminer ?

**Comment by hackmod** [comment]
as you can, see the seedhash is just fine.
~~~python
def get_seedhash(block):
     s = '\x00' * 32
     for i in range(block.number // EPOCH_LENGTH): # block.number is less then EPOCH_LENGTH in this case.
         s = serialize_hash(sha3_256(s))
     return s
~~~

> R u trying to implement ProgPoW on ethminer ?

noop. but just doing some test with the ProgPoW fork (by ifelsedef) on top of the latest ethminer.

**Comment by AndreaLanfranchi** [comment]
I think it should run **at least** one time as at block 0 we're anyhow in epoch 1

**Comment by AndreaLanfranchi** [comment]
I have amended the problem bad request when there is no path in the URI.

**Comment by AndreaLanfranchi** [comment]
@hackmod 

I was wrong about assumption of epoch 0.

Just pushed new commit in fork.
Please try and report

**Comment by hackmod** [comment]
works nicely!! (with or without "worker" parameter)

thank you! 👍

**Comment by AndreaLanfranchi** [comment]
Resolved by #1704
